Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: 
surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [1]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv', sep=',', header = None, 
                 names= ['surgery?', 'Age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 
                         'peripheral pulse', 'mucous membranes', 'capillary refill time', 'pain', 'peristalsis', 'abdominal distension', 
                         'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination', 'abdomen', 
                         'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome', 
                         'surgical lesion?', '25 type of lesion', '26 type of lesion', '27 type of lesion', 'cp_data'], na_values = '?')

df.head()

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0       2.0    1           530101                38.5   66.0   
1       1.0    1           534817                39.2   88.0   
2       2.0    1           530334                38.3   40.0   
3       1.0    9          5290409                39.1  164.0   
4       2.0    1           530255                37.3  104.0   

   respiratory rate  temperature of extremities  peripheral pulse  \
0              28.0                         3.0               3.0   
1              20.0                         NaN               NaN   
2              24.0                         1.0               1.0   
3              84.0                         4.0               1.0   
4              35.0                         NaN               NaN   

   mucous membranes  capillary refill time  ...  packed cell volume  \
0               NaN                    2.0  ...                45.0   
1               4.0                    1.0  ...                50.0   
2               3.0                    1.0  ...                33.0   
3               6.0                    2.0  ...                48.0   
4               6.0                    2.0  ...                74.0   

   total protein  abdominocentesis appearance  abdomcentesis total protein  \
0            8.4                          NaN                          NaN   
1           85.0                          2.0                          2.0   
2            6.7                          NaN                          NaN   
3            7.2                          3.0                          5.3   
4            7.4                          NaN                          NaN   

   outcome  surgical lesion?  25 type of lesion  26 type of lesion  \
0      2.0                 2              11300                  0   
1      3.0                 2               2208                  0   
2      1.0                 2                  0                  0   
3      2.0                 1               2208                  0   
4      2.0                 2               4300                  0   

   27 type of lesion  cp_data  
0                  0        2  
1                  0        2  
2                  0        1  
3                  0        1  
4                  0        2  

[5 rows x 28 columns]

In [2]:
horse_database = df[['surgery?', 'Age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']]
horse_database.loc[(horse_database.Age == 9), 'Age'] = 2 #Заменим значение "9" на "2" в столбце "Age"
horse_database

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    2                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                 NaN  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[300 rows x 8 columns]

Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

In [3]:
horse_database.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.080000           38.167917   71.913043   
std      0.490305    0.271746            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

  1:  surgery?
          1 = Yes, it had surgery
          2 = It was treated without surgery

  2:  Age
          1 = Adult horse
          2 = Young (< 6 months)

  4:  rectal temperature
          - linear
          - in degrees celsius.
          - An elevated temp may occur due to infection.
          - temperature may be reduced when the animal is in late shock
          - normal temp is 37.8
          - this parameter will usually change as the problem progresses
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock

  5:  pulse
          - linear
          - the heart rate in beats per minute
          - is a reflection of the heart condition: 30 -40 is normal for adults
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate

   6:  respiratory rate
          - linear
          - normal rate is 8 to 10
          - usefulness is doubtful due to the great fluctuations

  7:  temperature of extremities
          - a subjective indication of peripheral circulation
          - possible values:
               1 = Normal
               2 = Warm
               3 = Cool
               4 = Cold
          - cool to cold extremities indicate possible shock
          - hot extremities should correlate with an elevated rectal temp.

11: pain - a subjective judgement of the horse's pain level
          - possible values:
               1 = alert, no pain
               2 = depressed
               3 = intermittent mild pain
               4 = intermittent severe pain
               5 = continuous severe pain
          - should NOT be treated as a ordered or discrete variable!
          - In general, the more painful, the more likely it is to require
            surgery
          - prior treatment of pain may mask the pain level to some extent

23: outcome
          - what eventually happened to the horse?
          - possible values:
               1 = lived
               2 = died
               3 = was euthanized
  
  Качественные данные - surgery?,	Age,	temperature of extremities,	pain,	outcome. Они не подлежат сравнению. Судя по выводу функции describe() с учетом описания базы данных, выбросы в них отсутствуют.

In [4]:
print('Основные показатели:', horse_database[['rectal temperature', 'pulse', 'respiratory rate']].describe(),sep='\n', end='\n\n')
print('Мода:', horse_database[['rectal temperature', 'pulse', 'respiratory rate']].round().mode(), sep='\n', end='\n\n')
print('Медиана:', horse_database[['rectal temperature', 'pulse', 'respiratory rate']].median(), sep='\n', end='\n\n')
print('Дисперсия:', horse_database[['rectal temperature', 'pulse', 'respiratory rate']].var(), sep='\n', end='\n\n')

Основные показатели:
       rectal temperature       pulse  respiratory rate
count          240.000000  276.000000        242.000000
mean            38.167917   71.913043         30.417355
std              0.732289   28.630557         17.642231
min             35.400000   30.000000          8.000000
25%             37.800000   48.000000         18.500000
50%             38.200000   64.000000         24.500000
75%             38.500000   88.000000         36.000000
max             40.800000  184.000000         96.000000

Мода:
   rectal temperature  pulse  respiratory rate
0                38.0   48.0              20.0

Медиана:
rectal temperature    38.2
pulse                 64.0
respiratory rate      24.5
dtype: float64

Дисперсия:
rectal temperature      0.536247
pulse                 819.708775
respiratory rate      311.248328
dtype: float64



## Находим выбросы

In [5]:
q1_rt = horse_database['rectal temperature'].quantile(0.25)
q3_rt = horse_database['rectal temperature'].quantile(0.75)
iqr_rt = q3_rt - q1_rt
lower_bound_rt = q1_rt - (1.5 * iqr_rt) 
upper_bound_rt = q3_rt + (1.5 * iqr_rt)
remove_outliers_rt = horse_database[horse_database['rectal temperature'].between(lower_bound_rt, upper_bound_rt, inclusive=True)].sort_values('rectal temperature')
rt_table = pd.concat([horse_database, remove_outliers_rt]).drop_duplicates(keep=False)
list(rt_table['rectal temperature'].dropna())

[39.9,
 35.4,
 40.3,
 39.7,
 36.4,
 40.3,
 39.6,
 36.5,
 36.0,
 36.1,
 36.6,
 40.8,
 40.0,
 36.5]

In [6]:
q1_p = horse_database['pulse'].quantile(0.25)
q3_p = horse_database['pulse'].quantile(0.75)
iqr_p = q3_p - q1_p
lower_bound_p = q1_p - (1.5 * iqr_p) 
upper_bound_p = q3_p + (1.5 * iqr_p)
remove_outliers_p = horse_database[horse_database['pulse'].between(lower_bound_p, upper_bound_p, inclusive=True)].sort_values('pulse')
p_table = pd.concat([horse_database, remove_outliers_p]).drop_duplicates(keep=False)
list(p_table['pulse'].dropna())

[164.0, 150.0, 160.0, 184.0, 150.0]

In [7]:
q1_rr = horse_database['respiratory rate'].quantile(0.25)
q3_rr = horse_database['respiratory rate'].quantile(0.75)
iqr_rr = q3_rr - q1_rr
lower_bound_rr = q1_rr - (1.5 * iqr_rr) 
upper_bound_rr = q3_rr + (1.5 * iqr_rr)
remove_outliers_rr = horse_database[horse_database['respiratory rate'].between(lower_bound_rr, upper_bound_rr, inclusive=True)].sort_values('respiratory rate')
rr_table = pd.concat([horse_database, remove_outliers_rr]).drop_duplicates(keep=False)
list(rr_table['respiratory rate'].dropna())

[84.0,
 96.0,
 72.0,
 80.0,
 80.0,
 68.0,
 96.0,
 66.0,
 68.0,
 80.0,
 70.0,
 88.0,
 84.0,
 68.0,
 70.0]

В данном примере выбросы не являются аномальными, будем считать их вариантом нормы. Поэтому они не требуют удаления.

Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. 
Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. 
Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [8]:
print('Количество пропусков, шт. :', horse_database.isna().sum(), sep='\n', end='\n\n')
print('Количество пропусков, % :', (horse_database.isna().mean()*100).round(2), sep='\n')

Количество пропусков, шт. :
surgery?                       1
Age                            0
rectal temperature            60
pulse                         24
respiratory rate              58
temperature of extremities    56
pain                          55
outcome                        1
dtype: int64

Количество пропусков, % :
surgery?                       0.33
Age                            0.00
rectal temperature            20.00
pulse                          8.00
respiratory rate              19.33
temperature of extremities    18.67
pain                          18.33
outcome                        0.33
dtype: float64


Создаем датафрейм для заполнения пропусков:

In [9]:
test_fill = df[['surgery?', 'Age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']]
test_fill.loc[(horse_database.Age == 9), 'Age'] = 2

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [10]:
test_fill.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [11]:
test_fill['respiratory rate'].fillna(test_fill['respiratory rate'].mean(), inplace=True)
test_fill['pulse'].fillna(test_fill['pulse'].mean(), inplace=True)
test_fill['rectal temperature'].fillna(test_fill['rectal temperature'].mean(), inplace=True)
test_fill.describe()
# Заполнил количественные данные средними значениями, при этом статистические показатели изменились незначительно

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4463: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          300.000000  300.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.654705   27.457472   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.167917   68.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        300.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           15.838952                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           20.000000                    1.000000    2.00000    1.000000  
50%           30.000000                    3.000000    3.00000    1.000000  
75%           34.250000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [13]:
print(test_fill.groupby(['surgery?', 'outcome'])['pain'].median()) 

surgery?  outcome
1.0       1.0        3.0
          2.0        4.0
          3.0        4.0
2.0       1.0        2.0
          2.0        4.0
          3.0        3.0
Name: pain, dtype: float64


In [14]:
print(test_fill.groupby('pain')['temperature of extremities'].median())

pain
1.0    1.5
2.0    3.0
3.0    2.0
4.0    3.0
5.0    3.0
Name: temperature of extremities, dtype: float64


In [15]:
test_fill['pain'].fillna(test_fill.groupby(['surgery?', 'outcome'])['pain'].transform('median'), inplace=True)
test_fill['temperature of extremities'].fillna(test_fill.groupby('pain')['temperature of extremities']
                                               .transform('median'), inplace=True)

# Столбец 'pain'  в зависимости от того, была ли операция у лошади и каков исход(жива/мертва/усыплена)
# Столбец 'temperature of extremities' кореллирует в высокой температурой тела, которая зависит от боли.
# Поэтому заполнил медианным значением с группировкой по уровню боли.

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4463: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [16]:
test_fill.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          300.000000  300.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.654705   27.457472   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.167917   68.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        300.000000                  300.000000  300.000000  299.000000  
mean          30.417355                    2.400000    2.976667    1.551839  
std           15.838952                    0.973735    1.227931    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           20.000000                    1.000000    2.000000    1.000000  
50%           30.000000                    3.000000    3.000000    1.000000  
75%           34.250000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000